packages/fdc3-agent-proxy/src/channels/DefaultChannelSupport.ts

```diff
@@ -107,8 +131,12 @@ export class DefaultChannelSupport implements ChannelSupport {
       request,
       'createPrivateChannelResponse'
     );
-
-    return new DefaultPrivateChannel(this.messaging, response.payload?.privateChannel?.id!!);
+    if (response.payload.privateChannel) {
```

Comment:
ok, my other comment got lost, so re-creating it.

Problem is typescript is less discriminatory than JSON schema.  In the schema, if error is set then channel isn't and vice-versa.  If we get to this line, we know error wasn't set, so channel must be.

Therefore we have to decide:
  - If we assume the primacy of JSON schema, we should probably validate the message types in AbstractMessaging using @Roaders new functions and ! out the typescript error, or
  - if we say typescript is the golden source of truth, we need to add tests for all these mutant variations of JSON messages that would come through. 

Personally I think we should go with the former since we have those functions now.

Reply:
> The exchange function is testing for an error, and if that's not provided, the channel should be set, shouldn't it?

It should be, but thats relying on 3rd party code being implemented exactly correctly. If a DA developer fails to set either the error or payload elements correctly then they end up in the else clause and get both a specified error and log message they can hunt for. The types allow that as you could return either error or privateChannel so both end up optional.

> but since it's not reachable I don't think we can write one without passing in invalid JSON?

It is reachable with an implementation error in 3rd party code that may not have been written yet.

> Your JSON schema precludes this line of code from being reached?

Schema does, but types don't. Its just a defensive pattern that will better handle errors in 3rd party code.

> Perhaps a better way around this would be to make sure that the exchange function validates the JSON that comes back.

It would be possible to move it there, but either exchange will get much bigger, or it will need to start running validations on every response from a variety of schemas. Thats likely to affect performance at runtime. Whereas these checks on responses are light and easy to understand. I know you're not keen on the if/else statements due to test coverage - there's probably a way to write a general purpose validation function (perhaps not full schema validation, but the presence of a key property) to sort that out. That would look something like:

```ts
async createPrivateChannel(): Promise<PrivateChannel> {
    const request: CreatePrivateChannelRequest = {
      meta: this.messaging.createMeta(),
      type: 'createPrivateChannelRequest',
      payload: {},
    };
    const response = await this.messaging.exchange<CreatePrivateChannelResponse>(
      request,
      'createPrivateChannelResponse'
    );

    throwIfUndefined(response.payload.privateChannel, 'Invalid response from Desktop Agent to createPrivateChannel!',response,ChannelError.CreationFailed);
    
    return new DefaultPrivateChannel(this.messaging, response.payload.privateChannel!.id);
  }
```

Reply:
that might be a good approach.   One possible improvement might be to move the validation/throw into the `exchange` function as this is going to be needed a lot

Reply:
I thought about it - but it makes exchange longer and customizing the error/log message is harder there... I've installed that new approach whereever I can so that should cut down on uncovered log lines.

Reply:
(I'll push that up shortly)

Reply:
LGTM